core/trino-main/src/test/java/io/trino/operator/TestDeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,443 @@
+        }
+    }
+
+    private static void assertContainsPages(ExchangeClientBuffer buffer, List<String> values)
+    {
+        List<String> actual = new ArrayList<>();
+        while (true) {
+            SerializedPage page = buffer.pollPage();
+            if (page == null) {
+                break;
+            }
+            actual.add(page.getSlice().toStringUtf8());
+        }
+        assertTrue(buffer.isFinished());
+        assertThat(actual).containsExactlyInAnyOrderElementsOf(values);
+    }
+
+    @Test
+    public void testRemovePagesForPreviousAttempts()
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+
+            TaskId task1 = createTaskId(0, 0);
+            TaskId task2 = createTaskId(1, 0);
+            TaskId task3 = createTaskId(0, 1);
+
+            SerializedPage page1 = createPage("textofrandomlength");
+            SerializedPage page2 = createPage("textwithdifferentlength");
+            SerializedPage page3 = createPage("smalltext");
+
+            buffer.addTask(task1);
+            buffer.addPages(task1, ImmutableList.of(page1));
+            buffer.addTask(task2);
+            buffer.addPages(task2, ImmutableList.of(page2));
+
+            assertEquals(buffer.getRetainedSizeInBytes(), page1.getRetainedSizeInBytes() + page2.getRetainedSizeInBytes());
+
+            buffer.addTask(task3);
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+
+            buffer.addPages(task3, ImmutableList.of(page3));
+            assertEquals(buffer.getRetainedSizeInBytes(), page3.getRetainedSizeInBytes());
+        }
+    }
+
+    @Test
+    public void testBufferOverflow()
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), DataSize.of(100, BYTE), RetryPolicy.QUERY)) {
+            TaskId task = createTaskId(0, 0);
+
+            buffer.addTask(task);
+            SerializedPage page1 = createPage("1234");
+            buffer.addPages(task, ImmutableList.of(page1));
+
+            assertFalse(buffer.isFinished());
+            assertFalse(buffer.isBlocked().isDone());
+            assertEquals(buffer.getRetainedSizeInBytes(), page1.getRetainedSizeInBytes());
+
+            SerializedPage page2 = createPage("123456789");
+            buffer.addPages(task, ImmutableList.of(page2));
+            assertTrue(buffer.isFinished());
+            assertTrue(buffer.isBlocked().isDone());
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+            assertEquals(buffer.getBufferedPageCount(), 0);
+
+            assertThatThrownBy(buffer::pollPage)
+                    .isInstanceOf(TrinoException.class)
+                    .hasMessage("Retries for queries with large result set currently unsupported");
+        }
+    }
+
+    @Test
+    public void testIsFinished()
```

Comment:
For tests like this one, where we are testing multiple scenarios I would prefer to have a single line comment for each scenario explaining what we are testing in plain English:
// immediate close on an empty buffer
// empty set of tasks
// single task
...

Reply:
is this one addressed?

Reply:
I must've missed this. Let me add comments.